src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -85,66 +85,93 @@ namespace {
 
 #if GRPC_ENABLE_FORK_SUPPORT && GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK
 
+// Thread pool can outlive EE but we need to ensure the ordering if both
+// entities are alive.
+template <template <typename> typename PtrType>
+struct ForkHandlerPointers {
+  PtrType<PosixEventEngine> event_engine;
+  PtrType<ThreadPool> executor;
+
+  ForkHandlerPointers(PtrType<PosixEventEngine> ee,
+                      PtrType<ThreadPool> thread_pool)
+      : event_engine(std::move(ee)), executor(std::move(thread_pool)) {}
+};
+
 // Fork support - mutex and global list of event engines
 // Should never be destroyed to avoid race conditions on process shutdown
 absl::NoDestructor<grpc_core::Mutex> fork_mu;
-absl::NoDestructor<absl::InlinedVector<std::weak_ptr<PosixEventEngine>, 16>>
-    event_engines_for_fork_ ABSL_GUARDED_BY(fork_mu.get());
+absl::NoDestructor<absl::InlinedVector<ForkHandlerPointers<std::weak_ptr>, 16>>
+    fork_handlers ABSL_GUARDED_BY(fork_mu.get());
 
 // "Locks" event engines and returns a collection so callbacks can be invoked
 // without holding a lock.
-std::vector<std::shared_ptr<PosixEventEngine>> LockEventEngines() {
+std::vector<ForkHandlerPointers<std::shared_ptr>> LockForkHandlers() {
   grpc_core::MutexLock lock(fork_mu.get());
-  std::vector<std::shared_ptr<PosixEventEngine>> engines;
+  std::vector<ForkHandlerPointers<std::shared_ptr>> locked;
   // Not all weak_ptrs might be locked. If an engine enters dtor, it will stop
   // on a mutex in DeregisterEventEngineForFork but the weak pointer will not
   // be lockable here.
-  engines.reserve(event_engines_for_fork_->size());
-  for (const auto& engine : *event_engines_for_fork_) {
-    auto ptr = engine.lock();
-    if (ptr != nullptr) {
-      engines.emplace_back(std::move(ptr));
-    }
+  locked.reserve(fork_handlers->size());
+  for (const auto& [event_engine, executor] : *fork_handlers) {
+    locked.emplace_back(event_engine.lock(), executor.lock());
   }
-  return engines;
+  return locked;
 }
 
 void PrepareFork() {
-  for (const auto& engine : LockEventEngines()) {
-    engine->BeforeFork();
+  for (const auto& [event_engine, executor] : LockForkHandlers()) {
+    if (event_engine) {
```

Comment:
For pointer types, please prefer to say `!= nullptr` rather than evaluating them as booleans.

Same thing on throughout.

Reply:
Done.